L1Trigger/L1TTrackMatch/plugins/L1TkMuonProducer.cc

```diff
@@ -0,0 +1,712 @@
+  std::unique_ptr<TkMuonCollection> oc_tkmuon(new TkMuonCollection());
+  for (const auto& p : {oc_bmtf_tkmuon, oc_omtf_tkmuon, oc_emtf_tkmuon}) {
+    oc_tkmuon->insert(oc_tkmuon->end(), p.begin(), p.end());
+  }
+
+  // put the new track+muon objects in the event!
+  iEvent.put(std::move(oc_tkmuon));
+};
+
+void L1TkMuonProducer::runOnMTFCollection_v1(const edm::Handle<RegionalMuonCandBxCollection>& muonH,
+                                             const edm::Handle<L1TTTrackCollectionType>& l1tksH,
+                                             TkMuonCollection& tkMuons,
+                                             const int detector) const {
+  const L1TTTrackCollectionType& l1tks = (*l1tksH.product());
+  const RegionalMuonCandBxCollection& l1mtfs = (*muonH.product());
+
+  int imu = 0;
+  for (auto l1mu = l1mtfs.begin(0); l1mu != l1mtfs.end(0); ++l1mu) {  // considering BX = only
+
+    edm::Ref<RegionalMuonCandBxCollection> l1muRef(muonH, imu);
+    imu++;
+
+    float l1mu_eta = l1mu->hwEta() * 0.010875;
+    // get the global phi
+    float l1mu_phi = MicroGMTConfiguration::calcGlobalPhi(l1mu->hwPhi(), l1mu->trackFinderType(), l1mu->processor()) *
+                     2 * M_PI / 576.;
+
+    float l1mu_feta = fabs(l1mu_eta);
+    if (l1mu_feta < ETAMIN_)
+      continue;
+    if (l1mu_feta > ETAMAX_)
+      continue;
+
+    float drmin = 999;
+    float ptmax = -1;
+    if (ptmax < 0)
+      ptmax = -1;  // dummy
+
+    PropState matchProp;
+    int match_idx = -1;
+    int il1tk = -1;
+
+    int nTracksMatch = 0;
+
+    for (const auto& l1tk : l1tks) {
+      il1tk++;
+
+      float l1tk_pt = l1tk.momentum().perp();
+      if (l1tk_pt < PTMINTRA_)
+        continue;
+
+      float l1tk_z = l1tk.POCA().z();
+      if (fabs(l1tk_z) > ZMAX_)
+        continue;
+
+      float l1tk_chi2 = l1tk.chi2();
+      if (l1tk_chi2 > CHI2MAX_)
+        continue;
+
+      int l1tk_nstubs = l1tk.getStubRefs().size();
+      if (l1tk_nstubs < nStubsmin_)
+        continue;
+
+      float l1tk_eta = l1tk.momentum().eta();
+      float l1tk_phi = l1tk.momentum().phi();
+
+      float dr2 = deltaR2(l1mu_eta, l1mu_phi, l1tk_eta, l1tk_phi);
+      if (dr2 > 0.3)
+        continue;
+
+      nTracksMatch++;
+
+      const PropState& pstate = propagateToGMT(l1tk);
+      if (!pstate.valid)
+        continue;
+
+      float dr2prop = deltaR2(l1mu_eta, l1mu_phi, pstate.eta, pstate.phi);
+      // FIXME: check if this matching procedure can be improved with
+      // a pT dependent dR window
+      if (dr2prop < drmin) {
+        drmin = dr2prop;
+        match_idx = il1tk;
+        matchProp = pstate;
+      }
+    }  // over l1tks
+
+    LogDebug("MYDEBUG") << "matching index is " << match_idx;
+    if (match_idx >= 0) {
+      const L1TTTrackType& matchTk = l1tks[match_idx];
+
+      float sigmaEta = sigmaEtaTP(*l1mu);
+      float sigmaPhi = sigmaPhiTP(*l1mu);
+
+      float etaCut = 3. * sqrt(sigmaEta * sigmaEta + matchProp.sigmaEta * matchProp.sigmaEta);
```

Comment:
magic numbers should be named constants (everywhere)

Reply:
done